core/trino-main/src/main/java/io/trino/split/ConnectorAwareSplitSource.java

```diff
@@ -50,37 +69,59 @@ public CatalogHandle getCatalogHandle()
     @Override
     public ListenableFuture<SplitBatch> getNextBatch(int maxSize)
     {
+        checkState(source != null, "Already finished or closed");
         ListenableFuture<ConnectorSplitBatch> nextBatch = toListenableFuture(source.getNextBatch(maxSize));
         return Futures.transform(nextBatch, splitBatch -> {
             ImmutableList.Builder<Split> result = ImmutableList.builder();
             for (ConnectorSplit connectorSplit : splitBatch.getSplits()) {
                 result.add(new Split(catalogHandle, connectorSplit));
             }
-            return new SplitBatch(result.build(), splitBatch.isNoMoreSplits());
+            boolean noMoreSplits = splitBatch.isNoMoreSplits();
+            if (noMoreSplits) {
+                finished = true;
+                tableExecuteSplitsInfo = Optional.of(source.getTableExecuteSplitsInfo());
+                source = null;
```

Comment:
It seems like we're clearing the reference to `source` here without calling `source.close()`, which might cause problems in some of the underlying implementations. Should we just call `ConnectorAwareSplitSource#close()` here instead which will do both?

Reply:
thanks @pettyjamesm for https://github.com/trinodb/trino/pull/19109